### PR TITLE
Add tests for prepare_temp_table_for_import and insert_from_select

### LINKED ISSUE #4928: Add SQL tests for `prepare_temp_table_for_import` and `insert_from_select`
## Description

During the development of `bulk_insert` _aka_ Import into existing tables, we ran into bugs caused by some edge cases. Even though these bugs are now fixed, it would be good to have SQL tests for these so that we could avoid regressions in the future.

### Affects users

**None**

This may change if we do end up finding more bugs while writing tests.

### Affects developers

**Low**

Having these test would mean that we would avoid running into regressions in the future. This could also help us discover more bugs during the process of adding these tests.

## Suggested change

Tests to be added:
- For `msar.prepare_temp_table_for_import`
    - Passing one or more columns named `id` doesn't throw an error.
    - A `TEMPORARY TABLE` is actually created on the database with or without a given `tab_name`.
    - Happy case: Valid `tab_name` and `col_names` create a temporary table on the database with expected table and column names and return the `copy_sql` and `table_oid` for the relevant temporary table.
- For `msar.insert_from_select`
    - An error is thrown when inserting when mapping is made from a `text` type column into `integer` type column, i.e. there is a mismatch in datatypes of src and destination table and the src columns cannot be casted to valid dst column types.
    - Happy case: Records from the given `src` table _are_ inserted into `dst` table according to the `mappings` provided.

## Additional context

Feature implementation PRs: #4870, #4880

### PR DESCRIPTION
**Pull request by akankshahu**
Fixes #4928 

**Technical details**

Adds SQL tests for `msar.prepare_temp_table_for_import` and `msar.insert_from_select`.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by Anish9901 — CHANGES_REQUESTED** [review]
@akankshahu 

All SQL tests should go in `db/sql/test_sql_functions.sql` and must match the style of tests in that file.

Read the [DEVELOPER GUIDE](https://github.com/mathesar-foundation/mathesar/blob/develop/DEVELOPER_GUIDE.md) to know about how to run these tests locally.

**Comment by akankshahu** [comment]
@Anish9901  I have made some changes as per your request .Could  you please check?

File: db/sql/test_msar_insert_from_select.sql
Comment (Anish9901) [inline db/sql/test_msar_insert_from_select.sql:1]: This file shouldn't  exist anymore.

File: db/sql/test_msar_prepare_temp_table_for_import.sql
Comment (Anish9901) [inline db/sql/test_msar_prepare_temp_table_for_import.sql:1]: Same for this one. Shouldn't exist.

File: db/sql/test_sql_functions.sql [hunk 1]
```diff
@@ -1,7 +1,6 @@
 DROP EXTENSION IF EXISTS pgtap CASCADE;
 CREATE EXTENSION IF NOT EXISTS pgtap;
 
--- msar.drop_columns -------------------------------------------------------------------------------
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql:4]: Don't make changes unrelated to the issue.

File: db/sql/test_sql_functions.sql [hunk 2]
```diff
@@ -46,7 +45,6 @@ END;
 $$ LANGUAGE plpgsql;
 
 
--- msar.drop_table ---------------------------------------------------------------------------------
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql:49]: Same here, Don't remove this.

File: db/sql/test_sql_functions.sql [hunk 3]
```diff
@@ -298,6 +296,103 @@ END;
...
+CREATE OR REPLACE FUNCTION test_msar_prepare_temp_table_for_import() RETURNS SETOF TEXT AS $f$
+DECLARE
+  r_json jsonb;
+  table_oid bigint;
+BEGIN
+  PERFORM __setup_msar_prepare_temp_table();
+
+  -- Test with generated table name
+  r_json := msar.prepare_temp_table_for_import(
+    tab_name := 'tmp_test_' || replace(clock_timestamp()::text, ' ', '_'),
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Is there a specific reason why you decided to use `timestamps` for table names? You can use a regular name, the tables created exist only till the test is completed, once the test is complete, the tables are deleted.

File: db/sql/test_sql_functions.sql [hunk 4]
```diff
@@ -298,6 +296,103 @@ END;
...
+
+  r_json := msar.prepare_temp_table_for_import(
+    tab_name := 'tmp_copy_check_' || replace(clock_timestamp()::text, ' ', '_'),
+    col_names := ARRAY['id','other']
+  );
+  RETURN NEXT matches(
+    r_json->>'copy_sql',
+    '^COPY ',
+    'copy_sql returned and starts with COPY'
+  );
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Don't call `msar.prepare_temp_table_for_import` multiple times within the same test, call once and test all if returned json is as expected.

File: db/sql/test_sql_functions.sql [hunk 5]
```diff
@@ -298,6 +296,103 @@ END;
...
+$f$ LANGUAGE plpgsql;
+
+
+-- msar.insert_from_select -----------------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION __setup_msar_insert_from_select() RETURNS SETOF TEXT AS $$
+BEGIN
+  DROP TABLE IF EXISTS tmp_src_insert;
+  DROP TABLE IF EXISTS tmp_dst_insert;
+  CREATE TABLE tmp_src_insert (text_col text, num_col text);
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: This `src` table should be a `TEMPORARY` table.

File: db/sql/test_sql_functions.sql [hunk 6]
```diff
@@ -298,6 +296,103 @@ END;
...
+
+  RETURN NEXT throws_like($$
+    SELECT msar.insert_from_select(
+      'tmp_src_insert'::regclass,
+      'tmp_dst_insert'::regclass,
+      jsonb_build_array(
+        jsonb_build_object('src_table_attnum', 1, 'dst_table_attnum', 1)
+      )
+    );
+  $$, '%invalid input syntax for%integer%', 'insert_from_select raises when uncastable text -> integer mapping exists');
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: These should be split into two separate tests. One test shouldn't test multiple things.

File: db/sql/test_sql_functions.sql [hunk 7]
```diff
@@ -298,6 +296,103 @@ END;
 $f$ LANGUAGE plpgsql;
 
 
+
+CREATE OR REPLACE FUNCTION __setup_msar_prepare_temp_table() RETURNS SETOF TEXT AS $$
+BEGIN
+  DROP TABLE IF EXISTS tmp_src_prepare;
+  CREATE TABLE tmp_src_prepare(id text, other text);
+  INSERT INTO tmp_src_prepare VALUES ('1','a'), ('2','b');
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: These don't seem to be used anywhere.

File: db/sql/test_sql_functions.sql [hunk 8]
```diff
@@ -298,6 +296,103 @@ END;
...
+-- msar.insert_from_select -----------------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION __setup_msar_insert_from_select() RETURNS SETOF TEXT AS $$
+BEGIN
+  DROP TABLE IF EXISTS tmp_src_insert;
+  DROP TABLE IF EXISTS tmp_dst_insert;
+  CREATE TABLE tmp_src_insert (text_col text, num_col text);
+  -- Only castable data for successful insert tests
+  INSERT INTO tmp_src_insert VALUES ('100','10'), ('200','20');
+  CREATE TABLE tmp_dst_insert (id integer, value integer);
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Please add more columns with different types here, `insert_from_select()` needs to be tested to work on a wide variety of column types. Some of which include `Numeric`, `Boolean`, `Text`, `mathesar_types.email` and `mathesar_types.mathesar_money`.

Also, since the `dst` table is not a temporary table it shouldn't contain a `tmp_` prefix.

File: db/sql/test_sql_functions.sql [hunk 9]
```diff
@@ -298,6 +296,103 @@ END;
...
+END;
+$f$ LANGUAGE plpgsql;
+
+
+-- msar.insert_from_select -----------------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION __setup_msar_insert_from_select() RETURNS SETOF TEXT AS $$
+BEGIN
+  DROP TABLE IF EXISTS tmp_src_insert;
+  DROP TABLE IF EXISTS tmp_dst_insert;
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: These `DROP` commands aren't required, because tables are automatically cleaned up when they are created within the test via a `__setup`. So, you can call the same setup function in different tests without needing to worry about the same tables hanging around on the database.

**Review by Anish9901 — CHANGES_REQUESTED** [review]
Hey @akankshahu, This still need a lot of work to be mergeable. I've pointed out what needs to be changed.

**Comment by akankshahu** [comment]
@Anish9901  As you instructed, I have implemented the changes here. Could you please review it once?

File: db/sql/test_startup.sql [hunk 10]
```diff
@@ -8,6 +8,8 @@ $$ LANGUAGE plpgsql;
 CALL raise_notice('Creating testing DB');
 CREATE DATABASE mathesar_testing;
 \c mathesar_testing
+ALTER DATABASE mathesar_testing SET timezone TO 'UTC';
+SET TIME ZONE 'UTC';
```
Comment (Anish9901) [inline db/sql/test_startup.sql]: Not needed.

File: db/sql/test_sql_functions.sql [hunk 11]
```diff
@@ -298,6 +298,123 @@ END;
 $f$ LANGUAGE plpgsql;
 
 
+-- msar.prepare_temp_table_for_import -------------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION test_msar_prepare_temp_table_for_import() RETURNS SETOF TEXT AS $f$
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Please split this into multiple tests. Also, as mentioned in #4928, we need to test:

Test 1:
> Passing one or more columns named `id` doesn't throw an error.
- So we need to pass more than one column named `id` as `col_names` and check the column names created and make sure the returned json is as expected. 

Test 2:
> A `TEMPORARY TABLE` is actually created on the database with or without a given `tab_name`.

- You test _does_ include the case where the `tab_name` _is_ provided but not when `tab_name` isn't provided. Please add that.

Test 3:
> Happy case: Valid `tab_name` and `col_names` create a temporary table on the database with expected table and column names and return the `copy_sql` and `table_oid` for the relevant temporary table.
 
- Make sure you check the complete `COPY` statement and not whether is starts with `COPY`.

File: db/sql/test_sql_functions.sql [hunk 12]
```diff
@@ -298,6 +298,123 @@ END;
...
+    SELECT msar.insert_from_select(
+      'tmp_src_insert'::regclass,
+      'insert_dest_table'::regclass,
+      jsonb_build_array(
+        jsonb_build_object('src_table_attnum', 1, 'dst_table_attnum', 1),
+        jsonb_build_object('src_table_attnum', 2, 'dst_table_attnum', 2),
+        jsonb_build_object('src_table_attnum', 3, 'dst_table_attnum', 3),
+        jsonb_build_object('src_table_attnum', 4, 'dst_table_attnum', 4),
+        jsonb_build_object('src_table_attnum', 5, 'dst_table_attnum', 5),
+        jsonb_build_object('src_table_attnum', 6, 'dst_table_attnum', 6)
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Can you shuffle the `src` table columns such that the mappings aren't one-one i.e. make it so that e.g. column 3 of `src` goes to column 1 of `dst`, and so on.

File: db/sql/test_sql_functions.sql [hunk 13]
```diff
@@ -298,6 +298,123 @@ END;
...
+      )
+    );
+  $$, 'insert_from_select runs for castable mappings with various types');
+
+  RETURN NEXT is(
+    (SELECT COUNT(*)::bigint FROM insert_dest_table WHERE id IN (100,200)),
+    2::bigint,
+    'two rows inserted successfully'
+  );
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Check the actual records instead of just the count using `msar.list_records_from_table`. You can see how other tests in this file do it.

File: db/sql/test_sql_functions.sql [hunk 14]
```diff
@@ -298,6 +298,123 @@ END;
...
+
+  DROP TABLE insert_dest_table;
+END;
+$f$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION test_msar_insert_from_select_error() RETURNS SETOF TEXT AS $f$
+BEGIN
+  CREATE TEMPORARY TABLE tmp_src_error (text_col text, num_col text);
+  INSERT INTO tmp_src_error VALUES ('not_a_number','30');
+  CREATE TABLE insert_error_dest (id integer, value integer);
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Use `__setup_msar_insert_from_select()` instead.

**Review by Anish9901 — CHANGES_REQUESTED** [review]

File: db/sql/test_sql_functions.sql [hunk 15]
```diff
@@ -298,6 +298,233 @@ END;
...
+  
+  RETURN NEXT ok(
+    (r_json->>'table_oid')::bigint IS NOT NULL AND
+    (r_json->>'copy_sql') IS NOT NULL,
+    'returned json contains table_oid and copy_sql'
+  );
+END;
+$f$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION test_msar_prepare_temp_table_with_and_without_tab_name() RETURNS SETOF TEXT AS $f$
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Split into 2, 1 test for `with` and one for `without` tab name.

File: db/sql/test_sql_functions.sql [hunk 16]
```diff
@@ -298,6 +298,233 @@ END;
...
+  
+  RETURN NEXT is(
+    table_name,
+    'tmp_explicit_table',
+    'table created with specified name'
+  );
+  
+  -- Test with different tab_name to ensure temp table persistence
+  r_json := msar.prepare_temp_table_for_import(
+    tab_name := 'tmp_another_table',
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]: Without `tab_name` means this should be `null`.

File: db/sql/test_sql_functions.sql [hunk 17]
```diff
@@ -298,6 +298,233 @@ END;
...
+  );
+  
+  RETURN NEXT ok(
+    (r_json->>'table_oid')::bigint IS NOT NULL AND
+    (r_json->>'copy_sql') IS NOT NULL,
+    'returned json contains table_oid and copy_sql'
+  );
+END;
+$f$ LANGUAGE plpgsql;
+
```
Comment (Anish9901) [inline db/sql/test_sql_functions.sql]:
```suggestion


```

**Review by Anish9901 — APPROVED** [review]
Hey @akankshahu, Thanks for your efforts on this. Reviewing this PR was taking a lot of my time and the solution wasn't reaching to what was expected even after multiple rounds of reviews. So, I decided to fix it myself.

Please look through 93914cf, d78ee39 & cdc156c to see how my changes differ from yours.

Here are some pointers to keep in mind when you work on other issues:
- Read and understand the issue you are trying to fix carefully.
- LLMs make a lot of mistakes, make sure that you read through and understand the AI generated code & then check whether it is correct before you commit it.
- The PR descriptions is the first thing that any maintainer would see and it is a reflection of how much effort you've put into fixing the issue. So make sure it looks good and follows our PR template.
- I've fixed the description so that the checklist is correctly rendered, it should be `[x]` and not `[ x]` or `[x ]` notice how there is no white space before or after the `x`.

File: db/sql/05_msar.sql [hunk 18]
```diff
@@ -3419,7 +3419,7 @@ BEGIN
     END LOOP;
   END IF;
   column_defs := __msar.process_col_def_jsonb(0, col_defs, false);
-  PERFORM msar.add_temp_table(tab_name, column_defs);
+  PERFORM msar.add_temp_table(uq_tab_name, column_defs);
```
Comment (Anish9901) [inline db/sql/05_msar.sql:3422]: In the process of fixing these tests I was actually able to find a find a bug in the case when a `tab_name` isn't provided to `prepare_temp_table_for_import`, i.e. `tab_name` is `null`, this is fixed by passing `uq_tab_name` for `msar.add_temp_table`.
